mono/metadata/object.c

```diff
@@ -5305,44 +5313,46 @@ mono_runtime_try_invoke_array (MonoMethod *method, void *obj, MonoArray *params,
 	}
 }
 
-/**
- * mono_object_new:
- * \param klass the class of the object that we want to create
- * \returns a newly created object whose definition is
- * looked up using \p klass.   This will not invoke any constructors, 
- * so the consumer of this routine has to invoke any constructors on
- * its own to initialize the object.
- * 
- * It returns NULL on failure.
- */
+// FIXME these will move to header soon
+static  MonoObjectHandle
+mono_object_new_alloc_by_vtable (MonoVTable *vtable, MonoError *error);
+
+// FIXME these will move to header soon
+static MonoObjectHandle
+mono_object_new_by_vtable (MonoVTable *vtable, MonoError *error);
+
 MonoObject *
 mono_object_new (MonoDomain *domain, MonoClass *klass)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
-
+	HANDLE_FUNCTION_ENTER ();
 	ERROR_DECL (error);
-
-	MonoObject * result = mono_object_new_checked (domain, klass, error);
-
+	MonoObjectHandle o = mono_object_new_handle (domain, klass, error);
 	mono_error_cleanup (error);
-	return result;
+	HANDLE_FUNCTION_RETURN_OBJ (o);
 }
 
 MonoObject *
 ves_icall_object_new (MonoDomain *domain, MonoClass *klass)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
-
+	HANDLE_FUNCTION_ENTER ();
```

Comment:
could, in principle, use `ICALL_ENTRY()`/`ICALL_RETURN_OBJ` macros here.  We use them in a handlful of places in the runtime.

Reply:
Need to just convert it to a handles icall I think. A separate PR.

Reply:
That's actually a lot of work at the moment.

We have two sources of icalls:
1. the static table in icall-def.h
2. icalls added dynamically by `mono_register_jit_icall` 

For the former, but not the latter, we have support to automatically wrap them in handles.  The dynamic ones are the subject of https://github.com/mono/mono/issues/6850